components/planner/src/dynamo/planner/virtual_connector.py

```diff
@@ -0,0 +1,307 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+import asyncio
+import logging
+import time
+from typing import Optional
+
+from dynamo.planner.defaults import WORKER_COMPONENT_NAMES
+from dynamo.planner.planner_connector import PlannerConnector
+from dynamo.runtime import DistributedRuntime, EtcdKvCache
+from dynamo.runtime.logging import configure_dynamo_logging
+
+configure_dynamo_logging()
+logger = logging.getLogger(__name__)
+
+# Constants for scaling readiness check and waiting
+SCALING_CHECK_INTERVAL = 10  # Check every 10 seconds
+SCALING_MAX_WAIT_TIME = 1800  # Maximum wait time: 30 minutes (1800 seconds)
+SCALING_MAX_RETRIES = SCALING_MAX_WAIT_TIME // SCALING_CHECK_INTERVAL  # 180 retries
+
+
+class VirtualConnector(PlannerConnector):
+    """
+    This is a virtual connector for planner to output scaling decisions to non-native environments
+    This virtual connector does not actually scale the deployment, instead, it communicates with the non-native environment through ETCD
+    The deployment environment needs to read from ETCD to receive the scaling decisions and update ETCD to report scaling status
+    The prefix for the ETCD key is /{dynamo_namespace}/planner/
+    To output the scaling decisions, planner write to three keys:
+       - num_prefill_workers: an integer (stored as string), specifying how many prefill workers the deployment should have in the last scaling decision
+       - num_decode_workers: an integer (stored as string), specifying how many decode workers the deployment should have in the last scaling decision
+       - decision_id: an integer (stored as string), specifying an incremental id for the last scaling decision, if there's no scaling decision, the value would be -1
+    To receive the status of the scaling decisions, the deployment environment needs to write to this key:
+       - scaled_decision_id: an integer (stored as string), specifying if the newest decision_id that has been scaled
+    """
+
+    def __init__(
+        self, runtime: DistributedRuntime, dynamo_namespace: str, backend: str
+    ):
+        etcd_client = runtime.etcd_client()
+        if etcd_client is None:
+            raise RuntimeError("ETCD client is not initialized")
+
+        self.backend = backend
+        self.dynamo_namespace = dynamo_namespace
+        self.worker_component_names = WORKER_COMPONENT_NAMES[backend]
+
+        # Initialize current worker counts
+        self.num_prefill_workers = 0
+        self.num_decode_workers = 0
+        self.decision_id = -1
+
+        # Track when we first started skipping scaling due to unready state
+        self.first_skip_timestamp = None
+
+        # Store etcd_client for async initialization
+        self._etcd_client = etcd_client
+        self._etcd_kv_cache = None
```

Comment:
nit: we're not really storing the KV cache, so I think this should be called something more like `etcd_scaling_state`